src/Common/FileSegment.h

```diff
@@ -197,9 +210,17 @@ friend struct FileSegmentsHolder;
     size_t downloaded_size = 0;
     size_t reserved_size = 0;
 
+    /// global locking order rule:
+    /// 1. cache lock
+    /// 2. detach lock (if needed)
+    /// 3. segment lock
+
     mutable std::mutex mutex;
     std::condition_variable cv;
 
+    bool is_write_through_cache = false;
+    mutable std::mutex detach_mutex;
```

Comment:
Scope of this mutex is not clear to me. What it protects? Maybe we can just make some flags (is_detached and so on) atomic?